### PR TITLE
pkp/pkp-lib#427 Bulk editing for email templates

### PR DESCRIPTION
**Pull request by felixgruender**
Email templates can be exported as XML file and edited all at once.

### CODE REVIEW

File: pages/manager/EmailHandler.inc.php [hunk 1]
```diff
@@ -211,6 +211,197 @@ function enableEmail($args) {
...
+		if (!$fileName) {
+			$request->redirect(null, null, 'emails');
+		}
+		
+		$filesDir = Config::getVar('files', 'files_dir');
+		$filePath = $filesDir . '/journals/' . $journalId . '/' . $fileName;
+		
+		if (!$fileManager->uploadError($uploadName)) {
+			if ($fileManager->uploadedFileExists($uploadName)) {
+				if ($fileManager->uploadFile($uploadName, $filePath)) {
```
Comment (asmecher) [inline pages/manager/EmailHandler.inc.php]: Why not work with the file where it lives in the temporary upload area?

File: pages/manager/EmailHandler.inc.php [hunk 2]
```diff
@@ -211,6 +211,197 @@ function enableEmail($args) {
...
+	}
+	
+	/**
+	 * Check if an XML file contains valid XML
+	 * @param $filePath string
+	 * @return boolean
+	 */
+	function _isValidXML($filePath) {
+		if (is_file($filePath)) {
+			return is_object(simplexml_load_file($filePath));
```
Comment (asmecher) [inline pages/manager/EmailHandler.inc.php]: This results in double-parsing the file, and adds a simplexml dependency we don't have elsewhere.

**Comment by felixgruender** [comment]
I made the following 2 changes:
1) The XML file is checked now in the temporary upload directory (using FileManager::getUploadedFilePath()) - so the actual file upload isn't necessary any more
2) The _isValidXML() function has been removed - instead the result of XMLParser::parseStruct() is checked in the _saveEmailTemplates() function

File: pages/manager/EmailHandler.inc.php [hunk 3]
```diff
@@ -210,7 +210,188 @@ function enableEmail($args) {
...
 		Request::redirect(null, null, 'emails');
 	}
-
+	
+	/**
+	 * Export the selected email templates as XML
+	 * @param $args array
+	 * @@param $request PKPRequest
+	 */
+	function exportEmails($args, $request) {
```
Comment (asmecher) [inline pages/manager/EmailHandler.inc.php]: Be careful not to omit the calls to $this->validate() -- that's what checks that the user is a Journal Manager. Without it, anyone can access the new functions.

File: pages/manager/EmailHandler.inc.php [hunk 4]
```diff
@@ -210,7 +210,188 @@ function enableEmail($args) {
...
+		$filePath = $filesDir . '/journals/' . $journalId . '/' . $fileName;
+		
+		if (!$fileManager->uploadError($uploadName)) {
+			if ($fileManager->uploadedFileExists($uploadName)) {
+				$uploadedFilePath = $fileManager->getUploadedFilePath($uploadName);
+				if ($this->_saveEmailTemplates($uploadedFilePath, $journal)) {
+					if ($fileManager->deleteFile($uploadedFilePath)) {
+						$this->_showMessage($request);
+						$request->redirect(null, null, 'emails');
+						return true;
```
Comment (asmecher) [inline pages/manager/EmailHandler.inc.php]: I don't think the return values are ever used...

**Comment by felixgruender** [comment]
1) $this->validate() has been added to the new functions
2) superfluous return values have been removed in function uploadEmails()

**Pull request merged**